### PR TITLE
Change admission implementation

### PR DESCRIPTION
**Pull request by janboll**
This PR introduces change admission, by which change types can be declared as restrictive. Any changes to files covered by restrictive change types need to be done by approvers only. Changes could be allowed by adding a `/good-to-test` command though.
 
[APPSRE-8625](https://issues.redhat.com/browse/APPSRE-8625)

### CODE REVIEW

File: reconcile/change_owners/change_owners.py
Comment (patjlm) [inline reconcile/change_owners/change_owners.py:245]: should we allow anyone to submit, but need a `/good-to-test` handler of some sort?

Comment (janboll) [reply reconcile/change_owners/change_owners.py:245]: Yeah, I'd add that in a new iteration

File: reconcile/change_owners/change_owners.py
Comment (geoberle) [inline reconcile/change_owners/change_owners.py:257]: i think it would be better to signal failed assertion back to the caller and handle the exit there.
do we want to leave a message on the MR to help the author to figure out what is going on or is the output in the PR check logs enough?

at least we would need to mention who are the approvers and who would need to give their `/good-to-test`. can we add that somehow to the existing MR comment with all the approvers listed?

Comment (geoberle) [reply reconcile/change_owners/change_owners.py:257]: also, for MRs where AppSRE is the approriate approver i would love that the MR to shows up in the review queue

File: reconcile/gql_definitions/common/clusters.py
Comment (janboll) [reply reconcile/gql_definitions/common/clusters.py:648]: Might have had bad timing with generating the gql definitions. Let me rebase and do it again

File: reconcile/change_owners/change_owners.py
Comment (janboll) [reply reconcile/change_owners/change_owners.py:257]: Great feedback. I'll try to make it easier to digest for the user

**Comment by janboll** [comment]
Good to test label implemented/tested, reference: https://gitlab.cee.redhat.com/service/app-interface/-/merge_requests/92588

**Comment by janboll** [comment]
Review queue will be updated in another PR

File: reconcile/change_owners/change_owners.py
Comment (hemslo) [inline reconcile/change_owners/change_owners.py:266]: `user` here is the author, why check whether a change is admitted requires checking author? I thought the logic is just all restrictive changes have good to test approved. Like this:

```python
    return all(
        {a.org_username for a in c.approvers} & good_to_test_approvers
        for change in changes
        for dc in change.diff_coverage
        for c in dc.coverage
        if c.change_type_processor.restrictive
    )

```

Comment (janboll) [reply reconcile/change_owners/change_owners.py:266]: No it's more complex: Users with the restrictive role assigned are allowed to do these restrictive changes and add the looks good to test comment.

Comment (janboll) [reply reconcile/change_owners/change_owners.py:266]: A change is either admitted by schema or by a comment of an authorized person. I think the name is fine.

Comment (rporres) [reply reconcile/change_owners/change_owners.py:266]: What do you mean by "admitted by schema"? I don't really see where in this function only the schema check takes place (most likely is my newbiness to this integration).

I share a bit the feeling of not understanding well this function. The relationship between the `user`, the `approvers` and the `good_to_test_approvers` is complex.

File: reconcile/change_owners/change_owners.py
Comment (rporres) [inline reconcile/change_owners/change_owners.py:447]: We want to fail here because we assume this is the first integration that is going to run and we don't want to move on with the rest of the integrations, correct?

Comment (geoberle) [reply reconcile/change_owners/change_owners.py:447]: exactly

File: reconcile/change_owners/change_owners.py
Comment (hemslo) [reply reconcile/change_owners/change_owners.py:266]: I may understand what you mean. Unlike `lgtm` flow, even the author is part of approvers, still need a `lgtm` to mark it as approved. But for this admission check, if author is part of approvers, then it's admitted implicitly, right?

Btw this code is not covering all cases due to use of early `return`.

If the change has 2 different coverages, each one has a different list of approvers, one is `[A]`, the other one is `[B]`, only `A` commented good to test, then this change should be considered not admitted, require `B` to comment good to test.

To describe the condition:

* changes are admitted when all restrictive coverages are admitted
* a coverage is admitted when the author is part of approvers OR some approvers commented good to test

Translate to code:

```python

def is_coverage_admitted(coverage, user, good_to_test_approvers):
    return any(
        a.org_username == user or a.org_username in good_to_test_approvers
        for a in coverage.approvers
    )


def is_change_admitted(
    changes: list[BundleFileChange], user: str, good_to_test_approvers: set[str]
) -> bool:
    return all(
        is_coverage_admitted(c, user, good_to_test_approvers)
        for change in changes
        for dc in change.diff_coverage
        for c in dc.coverage
        if c.change_type_processor.restrictive
    )

```

Maybe should rename `user` to `author` to make it explicit.

File: reconcile/change_owners/change_owners.py
Comment (rporres) [reply reconcile/change_owners/change_owners.py:447]: Nice, thanks. A comment explaining this would be very helpful, especially because it express an expectation on how integrations are being run via hack scripts.

Comment (janboll) [reply reconcile/change_owners/change_owners.py:451]: added

File: reconcile/change_owners/change_owners.py
Comment (janboll) [reply reconcile/change_owners/change_owners.py]: updated

**Review by hemslo — APPROVED** [review]
LGTM!

**Review by rporres — APPROVED** [review]
nice!

**Pull request merged**
